Gems/Terrain/Code/Source/TerrainRenderer/TerrainMeshManager.cpp

```diff
@@ -64,31 +78,38 @@ namespace Terrain
         }
     }
 
+    void TerrainMeshManager::SetMaterial(MaterialInstance materialInstance)
+    {
+        if (m_materialInstance != materialInstance || m_materialInstance->GetCurrentChangeId() != m_lastMaterialChangeId)
+        {
+            m_lastMaterialChangeId = materialInstance->GetCurrentChangeId();
+            m_materialInstance = materialInstance;
+            m_forceRebuildDrawPackets = true;
+        }
+    }
+
     bool TerrainMeshManager::IsInitialized() const
     {
         return m_isInitialized;
     }
 
     void TerrainMeshManager::Reset()
     {
-        m_sectorModel = {};
         m_sectorStack.clear();
         m_rebuildSectors = true;
-        m_isInitialized = false;
     }
 
-    void TerrainMeshManager::Update(const AZ::RPI::ViewPtr mainView, AZ::Data::Instance<AZ::RPI::ShaderResourceGroup>& terrainSrg,
-        MaterialInstance materialInstance, AZ::RPI::Scene& parentScene, bool forceRebuildDrawPackets)
+    void TerrainMeshManager::Update(const AZ::RPI::ViewPtr mainView, AZ::Data::Instance<AZ::RPI::ShaderResourceGroup>& terrainSrg)
     {
         if (m_rebuildSectors)
         {
-            RebuildSectors(materialInstance, parentScene);
+            RebuildSectors();
             m_rebuildSectors = false;
         }
-        else if (forceRebuildDrawPackets)
+        else if (m_forceRebuildDrawPackets)
         {
             // The draw packets may need to be forcibly rebuilt in cases like a shader reload.
-            RebuildDrawPackets(parentScene);
+            RebuildDrawPackets();
         }
```

Comment:
Suggestion: If m_rebuildSectors and m_forceRebuildDrawPackets are meant to be exclusive (they can't both be true at the same time), you can enforce this constraint by creating a state enum instead of using two bools, which will make the chance of future errors lower.

Reply:
They can be true at the same time, so the actual fix is m_forceRebuildDrawPackets should be set to false inside the `if (m_rebuildSectors)` block. The logic is more like:
"if we have to rebuild all the sectors, then all the draw packets are going to get rebuilt anyway. If we just need to rebuild the draw packets (because say, a shader changed) then only do that work"